presto-hive-functions/src/main/java/com/facebook/presto/hive/functions/HiveFunctionNamespaceManager.java

```diff
@@ -0,0 +1,290 @@
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+public class HiveFunctionNamespaceManager
+        implements FunctionNamespaceManager<HiveFunction>, TypeManagerAware
+{
+    private static final long FUNCTION_NAME_EXPIRE_MILLS = 7_200_000;
+    public static final String ID = "hive-functions";
+
+    private final ClassLoader classLoader;
+    private final LoadingCache<FunctionKey, HiveFunction> functions;
+    private final LoadingCache<QualifiedObjectName, Boolean> functionNames;
+    private final HiveFunctionRegistry hiveFunctionRegistry;
+    private TypeManager typeManager;
+
+    @Inject
+    public HiveFunctionNamespaceManager(
+            @ForHiveFunction ClassLoader classLoader,
+            HiveFunctionRegistry hiveFunctionRegistry)
+    {
+        this.hiveFunctionRegistry = hiveFunctionRegistry;
+        this.classLoader = classLoader;
+        this.functions = CacheBuilder.newBuilder()
+                .build(CacheLoader.from(this::createFunction));
+        this.functionNames = CacheBuilder.newBuilder()
+                .expireAfterWrite(FUNCTION_NAME_EXPIRE_MILLS, TimeUnit.MILLISECONDS)
+                .build(CacheLoader.from(name -> {
+                    try {
+                        hiveFunctionRegistry.getClass(name);
+                    }
+                    catch (ClassNotFoundException e) {
+                        return false;
+                    }
+                    return true;
+                }));
+    }
+
+    @Override
+    public void setBlockEncodingSerde(BlockEncodingSerde blockEncodingSerde)
+    {
+        return;
+    }
+
+    public FunctionNamespaceTransactionHandle beginTransaction()
+    {
+        return new EmptyTransactionHandle();
+    }
+
+    public void commit(FunctionNamespaceTransactionHandle transactionHandle)
+    {
+    }
+
+    public void abort(FunctionNamespaceTransactionHandle transactionHandle)
+    {
+    }
+
+    public void createFunction(SqlInvokedFunction function, boolean replace)
+    {
+        throw new PrestoException(GENERIC_USER_ERROR, format("Cannot create function in hive function namespace: %s", function.getSignature().getName()));
+    }
+
+    public void alterFunction(QualifiedObjectName functionName, Optional<List<TypeSignature>> parameterTypes, AlterRoutineCharacteristics alterRoutineCharacteristics)
+    {
+        throw new PrestoException(GENERIC_USER_ERROR, format("Cannot alter function in hive function namespace: %s", functionName));
+    }
+
+    public void dropFunction(QualifiedObjectName functionName, Optional<List<TypeSignature>> parameterTypes, boolean exists)
+    {
+        throw new PrestoException(GENERIC_USER_ERROR, format("Cannot drop function in hive function namespace: %s", functionName));
+    }
+
+    @Override
+    public Collection<HiveFunction> listFunctions(Optional<String> likePattern, Optional<String> escape)
+    {
+        // TODO Implement list functions
```

Comment:
Shall we link a github issue for this TODO?

Reply:
Because we load the function during runtime. We can't know how much functions are supported after we support the external hive function.（some external hive functions may be loaded dynamically）
Now I list the funcitons that have been used before.